### PR TITLE
[tosa] Implement matmul, mm and bmm support

### PR DESCRIPTION
**Pull request by sjarus**
- Also handles braodcasting n-D tensors, dynamic shapes

Signed-off-by: Suraj Sudhir <suraj.sudhir@arm.com>

### CODE REVIEW

**Comment by sjarus** [comment]
Implements support for all matmul variants - matmul, mm and bmm. Supports 1D thru n-D, complex two-way broadcasting and dynamic shapes. 

The e2e tests that fail (e.g. Matmul_4d) do so because TOSA->LinAlg pass has an existing restriction around dynamic shapes and tosa.reshape() handling, which also affects several others, e.g. argmax, squeeze. 

Our internal unit tests (to be open sourced as part of Torch-MLIR CI) perform a number of permutations of matmul, generating Torch output, and TOSA bit level output from the TOSA reference model, and they match exactly: 

Results PASS test_matmul_16_x16__f32
Results PASS test_matmul_16_x16x32__f32
Results PASS test_matmul_16x32_x32__f32
Results PASS test_matmul_16x32_x32x17__f32
Results PASS test_matmul_8x16x32_x32__f32
Results PASS test_matmul_8x16x32_x32x17__f32
Results PASS test_matmul_16x32_x8x32x17__f32
Results PASS test_matmul_16_x8x16x32__f32
Results PASS test_matmul_8x16x32_x8x32x17__f32
Results PASS test_matmul_4x8x16x32_x32__f32
Results PASS test_matmul_4x8x16x32_x32x17__f32
Results PASS test_matmul_4x8x16x32_x8x32x17__f32
Results PASS test_matmul_4x8x16x32_x1x32x17__f32
Results PASS test_matmul_16_x4x8x16x32__f32
Results PASS test_matmul_4x8x16x32_x4x8x32x17__f32
Results PASS test_matmul_4x8x16x32_x1x8x32x17__f32
Results PASS test_matmul_4x8x16x32_x1x1x32x17__f32
Results PASS test_mm_16x32_x32x17__f32
Results PASS test_bmm_8x16x32_x8x32x17__f32
PASS: 211, FAIL: 0

**Review by cathyzhyi — COMMENTED** [review]
I haven't finished reviewing yet. Thanks a lot for the detailed comments! They are really helpful to understand the code.

One question I have is that seems we are not checking if the tensor shapes are legal, like if the matmul_dim are equal. The tensor shape requirements can also be different, some can only accept 2d/3d tensors. Should we check those as well?

**Comment by sjarus** [comment]
From my own experiments, torch.mm and torch.bmm require the right shaped tensors in order to process. Also, they emit failure if matrix-dim does not match. This is why I don't explicitly check that in this code - my understanding is that they will not parse and emit Torchscript form if these checks fail. However I can add those checks.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -678,6 +678,553 @@ LogicalResult ConvertAtenOp<AtenPowTensorScalarOp>::matchAndRewrite(
...
+                      rhsBroadcastedTy),
+                  rhs, rewriter.getI64ArrayAttr(rhsBroadcastedShape));
+
+    // TOSA matmul is performed on two 3D inputs and generates a 3D output.
+    // Lower ranked tensors are dim-1 reshaped up to 3D
+    auto reshapeUpTo3DTensor = [&](Value tensor) -> Value {
+      auto tensorTy = tensor.getType().cast<TensorType>();
+      auto rank = tensorTy.getRank();
+
+      assert(rank <= 3);
```
[the remark this answers is not in the document]
Comment (sjarus) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp]: Done

**Review by cathyzhyi — APPROVED** [review]
LGTM other than the runtime checks I mentioned here and in the previous comment.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -678,6 +678,553 @@ LogicalResult ConvertAtenOp<AtenPowTensorScalarOp>::matchAndRewrite(
...
+      // them back to their correct positions.
+
+      SmallVector<int64_t> transposedLhsShape;
+      SmallVector<int32_t> transposedLhsDims;
+
+      // Step: generate the common dim/shape information
+      for (uint32_t dim = 0; dim < maxInputRank - 2; dim++) {
+        bool isDynamicDim =
+            lhsBroadcastedTy.isDynamic(lhsBroadcastedShape[dim]);
+        if (isDynamicDim ||
```
[the remark this answers is not in the document]
Comment (sjarus) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:916]: Oh sorry didn't see this comment earlier. 

Dynamic shapes and broadcasting are a complicated topic. Examples:
1. how should a compiler handle broadcasting of 1 vs ? at runtime ? Should '?' be permitted to be any number, in which case the other dim must broadcast to it at runtime and the kernel should accommodate runtime broadcasting. Alternatively the compiler can assert '?' must be 1. 
2. How does runtime handle broadcast of '?' vs '?' ? Should they always match at runtime (no broadcast) ? Or can one of them be 1 (broadcast) ? 

In short, this would require the compiler stack to define these semantics, at which time it can enforce anything needed at compile time. It's for the above reasons that I haven't specifically sought to enforce something here. Hope this answers your question.

**Comment by sjarus** [comment]
> My understanding is that TorchScript can be generated without operands passed in. Once we got the compiled code, the illegal operands can be passed in to the compiled code and TorchScript is not involved at the runtime.

Right one can generate script-mode torchscript where operand shapes are not known - in this case checking is not possible. But if they are known then the trace mode TorchScript would fail if it could not execute and trace the output, wouldn't it ? E.g. 
>>> lhs = torch.randn((1,3,4))
>>> rhs = torch.randn((4,5))
>>> res = torch.mm(lhs, rhs)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: self must be a matrix

So this is why I didn't add any checks, but I suppose it doesn't hurt to add them.

**Comment by sjarus** [comment]
It's not clear what's change in TOSA->LinAlg but the previously working e2e tests fail during LinAlg legalization since the LLVM version just got updated. Is it ok for me to revert these tests from the passing lists and merge, and investigate that separately @cathyzhyi ?

**Comment by silvasean** [comment]
> Right one can generate script-mode torchscript where operand shapes are not known - in this case checking is not possible. But if they are known then the trace mode TorchScript would fail if it could not execute and trace the output, wouldn't it ? E.g.

They can be checked by emitting dynamic assertions into the lowered code. We do that in TorchToLinalg. For example, in [ConvertAtenMmOp](https://github.com/llvm/torch-mlir/blob/3745f54489ee30024bd55f3e95f86f1d24e9bd7e/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L970) we insert `tensor.dim` + `arith.cmpi` + `std.assert` ops to check that the requirements are met.

**Comment by cathyzhyi** [comment]
> It's not clear what's change in TOSA->LinAlg but the previously working e2e tests fail during LinAlg legalization since the LLVM version just got updated. Is it ok for me to revert these tests from the passing lists and merge, and investigate that separately @cathyzhyi ?

yea, that would be fine.

**Comment by sjarus** [comment]
> > Right one can generate script-mode torchscript where operand shapes are not known - in this case checking is not possible. But if they are known then the trace mode TorchScript would fail if it could not execute and trace the output, wouldn't it ? E.g.
> 
> They can be checked by emitting dynamic assertions into the lowered code. We do that in TorchToLinalg. For example, in [ConvertAtenMmOp](https://github.com/llvm/torch-mlir/blob/3745f54489ee30024bd55f3e95f86f1d24e9bd7e/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L970) we insert `tensor.dim` + `arith.cmpi` + `std.assert` ops to check that the requirements are met.

Given we're constructing an essentially TOSA-only form when doing TorchToTosa, these IR-expressed checks would sit more effectively at the TOSA->LinAlg interface the same way they're expressed when directly implementing Torch->LinAlg here; I'd certainly agree these are great checks to have at that abstraction.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (cathyzhyi) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:928]: yea, the compiler stack doesn't fully support broadcasting semantics now. In TorchToLinalg, we would avoid broadcasting for dynamic dims and and runtime assertion to enforce that limitation https://github.com/llvm/mlir-npcomp/blob/d6b6c0268cfd6ead0aa5f2f21ba02bbe8ead3180/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L2360-L2370
